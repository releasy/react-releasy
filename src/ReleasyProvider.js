// @flow
import * as React from 'react';
import * as PropTypes from 'prop-types';
import { Environment } from 'relay-runtime';
import invariant from 'invariant';

import { getEnvironment } from './store';

type Props = {
  children: React.Node,
};

class ReleasyProvider extends React.PureComponent<Props> {
  static propTypes = {
    children: PropTypes.element.isRequired,
  };

  static childContextTypes = {
    environment: PropTypes.instanceOf(Environment).isRequired,
  };

  constructor(props: Props) {
    invariant(props.children, 'ReleasyProvider: missing "children"');

    super(props);
  }

  getChildContext() {
    return {
      environment: getEnvironment(),
    };
  }

  render() {
    return this.props.children;
  }
}

export default ReleasyProvider;
