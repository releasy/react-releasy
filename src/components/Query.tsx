import * as React from 'react';
import { QueryRenderer } from 'react-relay';
import { GraphQLTaggedNode, Variables } from 'relay-runtime';

import GlobalContext, { GlobalContextValueType } from '../contexts/GlobalContext';

export type QueryChildrenPropsType = {
  error?: Error,
  retry: Function,
  isFetching: boolean,
};

type QueryPropsType = {
  query: GraphQLTaggedNode,
  variables?: Variables | Function,
  children: (props: QueryChildrenPropsType) => JSX.Element,
};

const Query = (props: QueryPropsType): React.ReactElement<any> => {
  const {
    query,
    variables,
    children,
  } = props;

  const sanitizedVariables = typeof variables === 'function' ? variables(props) : variables;

  return (
    <GlobalContext.Consumer>
      {({ environment }: GlobalContextValueType) =>
        <QueryRenderer
          environment={environment}
          query={query}
          variables={sanitizedVariables}
          render={({ props: queryProps, error, retry }) => {
            return children({
              ...queryProps,
              error,
              retry,
              isFetching: !queryProps && !error,
            });
          }}
        />
      }
    </GlobalContext.Consumer>
  );
};

export default Query;
