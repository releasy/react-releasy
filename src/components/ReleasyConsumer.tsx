import * as React from 'react';

import GlobalContext, { GlobalContextValueType } from '../contexts/GlobalContext';

type ReleasyConsumerPropsType = {
  children: React.ReactNode,
};

class ReleasyConsumer extends React.PureComponent<ReleasyConsumerPropsType> {
  render() {
    const { children } = this.props;

    return (
      <GlobalContext.Consumer>
        {(globalContextValue: GlobalContextValueType) =>
          React.Children.map(children, (child: React.ReactElement<any>) =>
            React.cloneElement(child, globalContextValue),
          )
        }
      </GlobalContext.Consumer>
    );
  }
}

export default ReleasyConsumer;
