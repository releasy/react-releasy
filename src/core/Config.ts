import { CacheConfig, RequestNode, UploadableMap, Variables, GraphQLResponse } from 'relay-runtime';
import invariant from 'invariant';

export interface LinkInterface {
  fetch: (
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables: UploadableMap,
  ) => Promise<GraphQLResponse>;
}

export interface CacheInterface {
  get: (queryID: string, variables: Variables) => GraphQLResponse;
  set: (queryID: string, variables: Variables, payload: GraphQLResponse) => void;
  clear: () => void;
  size: () => number;
}

type ConfigPropsType = {
  link: LinkInterface,
  cache?: CacheInterface,
  devTools?: boolean,
  networkLogger?: boolean,
};

class Config {
  link: LinkInterface;
  cache: CacheInterface;
  devTools: boolean;
  networkLogger: boolean;

  constructor(props: ConfigPropsType) {
    invariant(props, 'Config: missing props');
    invariant(props.link, 'Config: missing "link"');

    this.link = props.link;
    this.cache = props.cache || null;
    this.devTools = props.devTools || false;
    this.networkLogger = props.networkLogger || false;
  }
}

export default Config;
