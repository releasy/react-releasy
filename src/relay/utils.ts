import { RequestNode, CacheConfig, Variables } from 'relay-runtime';
import stableCopy from 'relay-runtime/lib/stableCopy.js';
import stringHash from 'string-hash';

export const isMutation = (request: RequestNode) => request.operationKind === 'mutation';

export const mustForceFetch = (cacheConfig: CacheConfig) => !!(cacheConfig && cacheConfig.force);

export const getHash = (request: RequestNode, variables: Variables): string => {
  const copy = stableCopy({
    variables: variables || {},
    query: request.text,
  });

  const { name } = request;
  const hash = stringHash(JSON.stringify(copy)).toString();

  return `${name}_${hash}`;
};
