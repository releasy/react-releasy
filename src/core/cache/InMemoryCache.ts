import { Variables, GraphQLResponse } from 'relay-runtime';
import RelayQueryResponseCache from 'relay-runtime/lib/RelayQueryResponseCache.js';

import { CacheInterface } from '../Config';

type InMemoryCacheProps = {
  ttl?: number,
  size?: number,
};

class InMemoryCache implements CacheInterface {
  store: RelayQueryResponseCache;

  constructor(props?: InMemoryCacheProps) {
    this.store = new RelayQueryResponseCache({
      size: props && props.size ? props.size : 250,
      ttl: props && props.ttl ? props.ttl : 60 * 1000 * 5, // default = 5 minutes
    });
  }

  get = (queryID: string, variables: Variables): GraphQLResponse => {
    return this.store.get(queryID, variables);
  }

  set = (queryID: string, variables: Variables, payload: GraphQLResponse): void => {
    this.store.set(queryID, variables, payload);
  }

  clear = (): void => {
    this.store.clear();
  }
}

export default InMemoryCache;
