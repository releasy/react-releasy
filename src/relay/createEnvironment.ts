import { installRelayDevTools } from 'relay-devtools';
import { Environment, RecordSource, Store } from 'relay-runtime';

import Config from '../core/Config';
import createNetwork from './createNetwork';

const createEnvironment = (config: Config): Environment => {
  if (config.devTools) {
    installRelayDevTools();
  }

  const network = createNetwork(config);

  const source = new RecordSource();
  const store = new Store(source);

  return new Environment({
    network,
    store,
  });
};

export default createEnvironment;
