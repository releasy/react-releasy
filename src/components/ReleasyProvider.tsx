import * as React from 'react';
import invariant from 'invariant';

import GlobalContext, {
  GlobalContextValueType,
  setConfig,
  setEnvironment,
} from '../contexts/GlobalContext';
import Config from '../core/Config';
import createEnvironment from '../relay/createEnvironment';

type ReleasyProviderPropsType = {
  config: Config,
  children: React.ReactNode,
};

type ReleasyProviderStateType = {
  globalContextValue: GlobalContextValueType,
};

class ReleasyProvider extends React.PureComponent<
  ReleasyProviderPropsType,
  ReleasyProviderStateType
> {
  constructor(props: ReleasyProviderPropsType) {
    super(props);

    invariant(props.config, 'ReleasyProvider: missing "config"');

    const { config } = props;
    const environment = createEnvironment(config);

    setConfig(config);
    setEnvironment(environment);

    this.state = {
      globalContextValue: {
        config,
        environment,
      },
    };
  }

  render() {
    const { children } = this.props;
    const { globalContextValue } = this.state;

    return (
      <GlobalContext.Provider value={globalContextValue}>
        {children}
      </GlobalContext.Provider>
    );
  }
}

export default ReleasyProvider;
