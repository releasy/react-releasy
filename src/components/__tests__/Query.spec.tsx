import * as React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { ChangeNameMutation, MeQuery } from '../../../tests/components';
import { createConfig } from '../../../tests/fixtures/core';
import { createQueryResponse, createMutationResponse } from '../../../tests/fixtures/relay';
import { delay } from '../../../tests/utils';

import InMemoryCache from '../../core/cache/InMemoryCache/InMemoryCache';
import ReleasyProvider from '../ReleasyProvider';

it('should render a loading Query', () => {
  global.fetch.mockResponseOnce(
    JSON.stringify(createQueryResponse()),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  const config = createConfig();

  const component = mount(
    <ReleasyProvider config={config}>
      <MeQuery />
    </ReleasyProvider>,
  );

  component.update();

  expect(toJson(component.find(MeQuery).find('div'))).toMatchSnapshot();
});

it('should render a success Query', async () => {
  global.fetch.mockResponseOnce(
    JSON.stringify(createQueryResponse()),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  const config = createConfig();

  const component = mount(
    <ReleasyProvider config={config}>
      <MeQuery />
    </ReleasyProvider>,
  );

  await delay(100);

  component.update();

  expect(toJson(component.find(MeQuery).find('div'))).toMatchSnapshot();
});

it('should use a function to create variables', async () => {
  global.fetch.mockResponseOnce(
    JSON.stringify(createQueryResponse()),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  const config = createConfig();

  const variables = () => ({
    customVariable: 'customValue',
  });

  const component = mount(
    <ReleasyProvider config={config}>
      <MeQuery variables={variables} />
    </ReleasyProvider>,
  );

  await delay(100);

  component.update();

  expect(toJson(component.find(MeQuery).find('div'))).toMatchSnapshot();
});

it('should store data in the cache', async () => {
  global.fetch.mockResponseOnce(
    JSON.stringify(createQueryResponse()),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  const cache = new InMemoryCache();
  const config = createConfig({ cache });

  mount(
    <ReleasyProvider config={config}>
      <MeQuery />
    </ReleasyProvider>,
  );

  await delay(100);

  expect(config.cache.size()).toBe(1);
});

it('should hit cache', async () => {
  global.fetch.mockResponseOnce(
    JSON.stringify(createQueryResponse()),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  const cache = new InMemoryCache();
  const config = createConfig({ cache });

  mount(
    <ReleasyProvider config={config}>
      <MeQuery />
    </ReleasyProvider>,
  );

  await delay(100);

  mount(
    <ReleasyProvider config={config}>
      <MeQuery />
    </ReleasyProvider>,
  );

  await delay(100);

  expect(config.cache.size()).toBe(1);
});

it('should clear cache on mutations', async () => {
  global.fetch.mockResponseOnce(
    JSON.stringify(createQueryResponse()),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  const cache = new InMemoryCache();
  const config = createConfig({ cache });

  mount(
    <ReleasyProvider config={config}>
      <MeQuery />
    </ReleasyProvider>,
  );

  await delay(100);

  global.fetch.mockResponseOnce(
    JSON.stringify(createMutationResponse()),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  mount(
    <ReleasyProvider config={config}>
      <ChangeNameMutation />
    </ReleasyProvider>,
  );

  await delay(100);

  expect(config.cache.size()).toBe(0);
});
