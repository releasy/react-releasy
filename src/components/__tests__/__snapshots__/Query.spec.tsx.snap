// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a loading Query 1`] = `
<div>
  loading
</div>
`;

exports[`should render a success Query 1`] = `
<div>
  fakeViewerId
  : 
  John Doe
</div>
`;

exports[`should use a function to create variables 1`] = `
<div>
  fakeViewerId
  : 
  John Doe
</div>
`;
